### PR TITLE
WIP: Add Build instructions for compiling a 64-Bit kernel

### PR DESCRIPTION
**Pull request by RyanHir**
In this merge request I will be adding instructions for compiling a 64-bit kernel for any device that supports the 64-bit instructions.

As the Raspberry Pi OS moves to be cross arch we should update the build instructions for the kernel to better clarify how to compile for both instruction sets. 

As the build instructions are mostly the same between arches, I will have notes in the markdown that states if X is specific to 32 or 64 bit.

This request is the predecessor of another pull request #1652 that I have done which I should have waited to do.

Tested build on 64-Bit Pi 3 and derivatives
Tested build on 64-Bit Pi 4

### CODE REVIEW

**Comment by RyanHir** [comment]
This request is the predecessor of another pull request #1652 that I have done which I should have waited to do.

**Comment by lurch** [comment]
Maybe this is a bit premature, and we ought to avoid adding this to the official documentation until the 64-bit Raspberry Pi OS is out of beta? :shrug:

**Comment by RyanHir** [comment]
That is reasonable, I will leave this pull request up until the 64-Bit Raspberry Pi OS comes out of beta for possible merging into the docs.

**Comment by lurch** [comment]
I've labelled it as WIP, to tag that it's not ready to be merged yet.

**Comment by JamesH65** [comment]
@popcornmix @pelwell Can either of you guys give this a quick once over to check its sane? Seems like a good addition to me.

**Comment by popcornmix** [comment]
Looks reasonable. I notice that "64-Bit Configs" section says "For Pi 3, Pi 3+ or Compute Module 3" and "For Raspberry Pi 4" with the less straightforward Raspberry Pi 2 ignored.

File: linux/kernel/building.md [hunk 1]
```diff
@@ -208,6 +251,17 @@ sudo umount mnt/fat32
...
 ```
 
+#### For 64-Bit
+
+```bash
+sudo cp mnt/fat32/$KERNEL.img mnt/fat32/$KERNEL-backup.img
+sudo cp arch/arm64/boot/Image mnt/fat32/$KERNEL.img
+sudo cp arch/arm64/boot/dts/broadcom/*.dtb mnt/fat32/
+sudo cp arch/arm64/boot/dts/overlays/*.dtb* mnt/fat32/overlays/
+sudo umount mnt/fat32
```
Comment (pelwell) [inline linux/kernel/building.md:261]: This fails to copy the README.

File: linux/kernel/building.md [hunk 2]
```diff
@@ -103,10 +103,21 @@ You can either do this using VirtualBox (or VMWare) on Windows, or install it di
 
 To build the sources for cross-compilation, make sure you have the dependencies needed on your machine by executing:
 ```bash
-sudo apt install git bc bison flex libssl-dev make libc6-dev libncurses5-dev crossbuild-essential-armhf
+sudo apt install git bc bison flex libssl-dev make libc6-dev libncurses5-dev
 ```
+
 If you find you need other things, please submit a pull request to change the documentation.
 
+#### Install the 32-Bit toolchain for a 32-Bit Kernel
```
Comment (pelwell) [inline linux/kernel/building.md:111]: On a stylistic note, I don't like the random capitalisation of certain words such as "Bit", "Kernel" and "Config".

File: linux/kernel/building.md [hunk 3]
```diff
@@ -92,7 +92,7 @@ sudo cp arch/arm/boot/zImage /boot/$KERNEL.img
 
 **Note**: On a Raspberry Pi 2/3/4, the `-j4` flag splits the work between all four cores, speeding up compilation significantly.
 
-## Cross-compiling
+## Cross-compiling 32/64-bit
```
Comment (pelwell) [inline linux/kernel/building.md:95]: Does this add anything?

File: linux/kernel/building.md [hunk 4]
```diff
@@ -190,14 +224,23 @@ sudo mount /dev/sdb6 mnt/fat32
...
-Next, install the modules:
 
+Next, install the kernel modules onto the SD card:
+
+#### For 32-Bit
 ```bash
 sudo env PATH=$PATH make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- INSTALL_MOD_PATH=mnt/ext4 modules_install
 ```
 
+#### For 64-Bit:
```
Comment (pelwell) [inline linux/kernel/building.md:235]: Inconsistent use of a colon.

File: linux/kernel/building.md (hunk 1 above)
Comment (RyanHir) [reply linux/kernel/building.md:261]: The readme is not generated on a his build so there is no file to copy

File: linux/kernel/building.md (hunk 4 above)
Comment (RyanHir) [reply linux/kernel/building.md]: Resolved in new commit

File: linux/kernel/building.md (hunk 3 above)
Comment (RyanHir) [reply linux/kernel/building.md]: Resolved in new commit

**Comment by RyanHir** [comment]
@popcornmix that change to add pi 2 instructions would not be possible for a 64-bit kernel as the device only has a 32-bit processor

File: linux/kernel/building.md (hunk 2 above)
Comment (RyanHir) [reply linux/kernel/building.md]: Resolved in b66bbe7

File: linux/kernel/building.md (hunk 1 above)
Comment (pelwell) [reply linux/kernel/building.md:261]: What do you mean not generated in this build? If you are only updating the 64-bit kernel then the README should be copied.

Comment (RyanHir) [reply linux/kernel/building.md:261]: When I build the kernel with the 64bit instructions, there is no README file to copy like the 32bit kernel build. I cant add to the instruction to copy a file that does not exist..

Comment (pelwell) [reply linux/kernel/building.md:261]: The overlays are built _from arch/arm_, and the README must be copied _from arch/arm_. Look at the 32-bit instructions and you might understand.

Comment (RyanHir) [reply linux/kernel/building.md:261]: interesting, i will do another build and check if it exists. If so, i will add it to the build instructions

Comment (pelwell) [reply linux/kernel/building.md:261]: It's a source file - it's just copied.

**Comment by popcornmix** [comment]
Newer revisions of Pi 2 are actually 64-bit (they use the same processor as Pi 3).
You can identify if this is the case from /proc/cpuinfo. But this does complicate the instructions so could be glossed over.

**Comment by RyanHir** [comment]
Being how small I assume the pool of pi 2 users interested in the 64-bit kernel, I don't think it is worth adding to the instructions.

On top of that, the Pi 2 is not supported for the 64-bit Raspberry Pi OS.

But thank you for this info, I was not aware some Pi 2's shipped with a Pi 3 processor.

File: linux/kernel/building.md (hunk 1 above)
Comment (RyanHir) [reply linux/kernel/building.md:262]: You are right, i was normally building into a separate folder so i got confused.

Fixed in f9e1697

File: linux/kernel/building.md [hunk 5]
```diff
@@ -103,10 +103,21 @@ You can either do this using VirtualBox (or VMWare) on Windows, or install it di
...
 ```
+
 If you find you need other things, please submit a pull request to change the documentation.
 
+#### Install the 32-bit toolchain for a 32-bit kernel
+```bash
+sudo apt install crossbuild-essential-armhf
+```
+
+#### Or, Install the 64-bit toolchain for a 64-bit kernel
```
Comment (lurch) [inline linux/kernel/building.md:116]: Lowercase i on 'install' here please

**Comment by lurch** [comment]
This only seems to add the cross-compiling instructions for building a 64-bit kernel. Should it also add the instructions for native-compiling a 64-bit kernel? (which I guess will be slightly different depending on whether you're running the 32-bit or 64-bit version of Raspberry Pi OS?)

File: linux/kernel/building.md (hunk 5 above)
Comment (RyanHir) [reply linux/kernel/building.md]: Fixed in 9ba510a

**Comment by RyanHir** [comment]
Adding instructions for compiling 32 -> 32 or 64 -> 64 is quite simple, but in the case of compiler from 32 -> 64 or 64 -> 32, you might as well be using the cross compile instructions as they are not arch specific.

I can add a clarifier to choose the correct configs for compiling a 64 -> 64 build.

File: linux/kernel/building.md [hunk 6]
```diff
@@ -103,10 +103,21 @@ You can either do this using VirtualBox (or VMWare) on Windows, or install it di
 
 To build the sources for cross-compilation, make sure you have the dependencies needed on your machine by executing:
 ```bash
-sudo apt install git bc bison flex libssl-dev make libc6-dev libncurses5-dev crossbuild-essential-armhf
+sudo apt install git bc bison flex libssl-dev make libc6-dev libncurses5-dev
 ```
+
 If you find you need other things, please submit a pull request to change the documentation.
 
+#### install the 32-bit toolchain for a 32-bit kernel
```
Comment (lurch) [inline linux/kernel/building.md:111]: We **do** want a capital `I` here, because it's the first word in the header :smile:

Comment (RyanHir) [reply linux/kernel/building.md]: My bad, miss understood your request. Fixed in f32ba83

File: linux/kernel/building.md [hunk 7]
```diff
@@ -208,6 +251,18 @@ sudo umount mnt/fat32
 sudo umount mnt/ext4
```
Comment (lurch) [inline linux/kernel/building.md:251]: I wonder if it might make sense to move the `sudo umount mnt/ext4` lines immediately after the `modules_install` lines? :shrug:

Comment (RyanHir) [reply linux/kernel/building.md:251]: While i like that idea, i think that may be outside the scope of this PR, if you fell otherwise I would be glad to change it.

Comment (pelwell) [reply linux/kernel/building.md:251]: I agree - let's not make the perfect the enemy of the good.

**Comment by lurch** [comment]
>  but in the case of compiler from 32 -> 64 or 64 -> 32, you might as well be using the cross compile instructions as they are not arch specific.

Probably worth adding a note to that effect? Something like "If you want to compile a 64-bit kernel on a 32-bit Raspberry Pi OS, or compile a 32-bit kernel on a 64-bit Raspberry Pi OS, then follow the cross-compilation instructions elsewhere on this page." ?
Or do you think that would be too confusing @pelwell @JamesH65 ?

> I can add a clarifier to choose the correct configs for compiling a 64 -> 64 build.

I think that would be useful :+1: 
(as well as perhaps clarifying that the existing instructions are for compiling a 32-bit kernel on a 32-bit Raspberry Pi OS?)

**Comment by pelwell** [comment]
It's not confusing, it's correct. What's needed is a small note near the top of the file saying briefly what cross-compilation is (building for an architecture that isn't the one doing the building), including 64-bit on 32-bit and vice versa as example, and referring to the cross-compilation section.

**Review by RyanHir — COMMENTED** [review]
EDIT: ~~This probably needs some updating but it is a start~~ This was meant to be a comment on a individual patch, look ahead.

File: linux/kernel/building.md [hunk 8]
```diff
@@ -1,8 +1,8 @@
 # Kernel building
 
-There are two main methods for building the kernel. You can build locally on a Raspberry Pi, which will take a long time; or you can cross-compile, which is much quicker, but requires more setup.
+There are two ways of compiling the linux kernel. One is to compile nativly or cross architecture via cross compiling. Cross compiling is used to compile the kernel for an architecture (such as arm64) on when on another architecture (such as armhf on the Raspberry Pi OS 32-bit).
```
Comment (RyanHir) [inline linux/kernel/building.md:3]: I meant this paragraph probably needs updating but I added some starting information

Comment (lurch) [reply linux/kernel/building.md:3]: Something like this might be better?

There are two main methods for building the kernel - natively and cross-compiling. Native compilation is the easiest approach, and is used when the architecture you're building the kernel on (host platform) is the same as the architecture you'll be running it on (target platform); such as building a 32-bit Raspberry Pi kernel using 32-bit Raspberry Pi OS, or building a 64-bit Raspberry Pi kernel using 64-bit Raspberry Pi OS. Cross-compiling is used when the "host platform" is different to the "target platform", such as building a 64-bit Raspberry Pi kernel using 32-bit Raspberry Pi OS (or vice versa), or when building a Raspberry Pi kernel (either 32 or 64-bit) on an separate Linux PC. Cross-compilation is slightly more involved, but building a Raspberry Pi kernel is typically much faster when cross-compiled on a separate Linux PC than when built natively on a Raspberry Pi.

I suspect @pelwell and @JamesH65 may have other suggestions / tweaks to the above too. Not sure if what I've suggested here is too wordy?

File: linux/kernel/configuring.md [hunk 9]
```diff
@@ -22,12 +22,18 @@ Once you've got everything set up and ready to go, you can compile and run the `
 $ make menuconfig
 ```
 
-If you're cross-compiling,:
+If you're cross-compiling a 32-Bit kernel:
```
Comment (lurch) [inline linux/kernel/configuring.md:25]: Lower-case 'b' on 'bit' here (and also below) please :slightly_smiling_face:

File: linux/kernel/building.md (hunk 8 above)
Comment (pelwell) [reply linux/kernel/building.md:3]: My attempt:
> The default compilers and linkers that come with an OS are configured to build executables to run on that OS - they are _native_ tools - but that doesn't have to be the case. A cross-compiler is configured to build code for a target other than the one running the build process, and using it is called cross-compilation.
>
> Cross-compilation of the Raspberry Pi kernel is useful for two reasons:
> 1. it allows a 64-bit kernel to be built using a 32-bit OS, and vice versa, and
> 2. even a modest laptop can cross-compile a Pi kernel significantly faster than the Pi itself.
>
> The instructions below are divided into native builds and cross-compilation; choose the section appropriate for your situation - although there are many common steps between the two there are also some important differences.

File: linux/kernel/configuring.md (hunk 9 above)
Comment (RyanHir) [reply linux/kernel/configuring.md]: Fixed in 4ee1b6a

File: linux/kernel/building.md (hunk 8 above)
Comment (RyanHir) [reply linux/kernel/building.md]: I personally like @pelwell 's attempt as it is not as wordy. Perhaps we should add hyper links to the section of the document in the header?

Comment (RyanHir) [reply linux/kernel/building.md]: I will add pelwell's header for now

**Comment by pelwell** [comment]
Is there anything outstanding on this PR before it can be merged? We've recently pointed people at this PR because it's better than the current published version.

**Comment by lurch** [comment]
I think it still needs "compiling a 64-bit kernel on 64-bit Raspberry Pi OS" instructions?

**Comment by pelwell** [comment]
We've not had that before - can't it be added in a follow-up commit?

**Comment by lurch** [comment]
I guess so! Although in that case I guess this PR ought to be renamed "Add Build instructions for cross-compiling a 64-bit kernel" :wink: 

This PR LGTM otherwise - thanks for your work on this @RyanHir
